[2201.9.0] Update openAPI learn page (#8981)
ballerina-platform/ballerina-dev-website: swan-lake/integration-tools/openapi-tool.md

>**Info:** For more command options, see [OpenAPI to Ballerina CLI options](#openapi-to-ballerina-command-options).

> ### Automate client generation with Package Build

Reviewer: The automation is supported for all the ballerina commands right? - `build`, `pack`, `test`

How about `Automate client generation using the Ballerina tool integration`?

Author: Yeah, `build, `pack`, `test` and  `run` commands execute at least some phases of the build process.  That is why here it used `package build`. 
@azinneera , @Dilhasha  will you be able to provide some input for the title we need to add for these tool scenarios?

Reviewer 2: Shall we say just "Automate client generation"?

Author: I will add it as the title

#### Update the Ballerina.toml file with OpenAPI tool configurations

> Adding the following OpenAPI tool configurations to the Ballerina.toml file will generate a client during the `bal build`. This eliminates the need to commit the generated code.

Reviewer: ```suggestion
Adding the following OpenAPI tool configurations to the `Ballerina.toml` file will generate a client during the `bal build`. This eliminates the need to commit the generated code.
```

> Adding the following OpenAPI tool configurations to the Ballerina.toml file will generate a client during the `bal build`. This eliminates the need to commit the generated code.

Reviewer: Here also we are only mentioning about `bal build` right?

> > **Info:** It is mandatory to provide `id`,` filePath`, and `targetModule` attributes for the OpenAPI tool configurations. The other attributes are optional.

Reviewer: The `targetModule` is also optional and defaulted to the `default` module. As a best practice we need to provide the `targetModule` otherwise it may have some conflicts to the files which are in the `default` module (outside the generated directory). Is that the reason for mentioning it as a `mandatory` option?

Author: will change this to optional.

> The following is an example of the tool configuration usage in the Ballerina.toml.

Reviewer: ```suggestion
The following is an example of the tool configuration usage in the `Ballerina.toml`.
```

| `options.statusCodeBinding`| This option can be used in the client generation to generate the client methods with status code response binding.                                                                                                                                                                                                                                                                                                                                                                                           |  Optional         |

#### Update the Ballerina.toml file with OpenAPI tool configurations using OpenAPI CLI command

> Executing the following OpenAPI `add` sub-command, along with the OpenAPI to Ballerina CLI options, will update the Ballerina.toml with OpenAPI tool configurations.

Reviewer: ```suggestion
Executing the following OpenAPI `add` sub-command, along with the OpenAPI to Ballerina CLI options, will update the `Ballerina.toml` with OpenAPI tool configurations.
```

>**Info:** For more command options, see [OpenAPI to Ballerina CLI options](#openapi-to-ballerina-command-options).

### Automate client generation

#### Update the `Ballerina.toml` file with OpenAPI tool configurations

Adding the following OpenAPI tool configurations to the `Ballerina.toml` file will generate a client during the package build. This eliminates the need to commit the generated code.

> > **Info:** It is mandatory to provide `id`, and ` filePath` attributes for the OpenAPI tool configurations. The other attributes are optional.

Reviewer 3: ```suggestion
> **Info:** It is mandatory to provide `id` and `filePath` attributes for the OpenAPI tool configurations. The other attributes are optional.
```

The following is an example of the tool configuration usage in the `Ballerina.toml`.

```ballerina
[[tool.openapi]]
id = "client01"
filePath = "./openapi.yaml"
targetModule = "delivery"
options.mode = "client"
options.statusCodeBinding = true
options.tags = ["tag1", "tag2"]
options.operations = ["op1", "op2"]
options.nullable = true # (default value => false)
options.license = "./license.txt"
```

> The below tool configuration can be used

Reviewer 3: ```suggestion
The below tool configuration can be used.
```

|----------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------|
| `filePath`   | The `filePath` tool option specifies the path of the OpenAPI contract file (e.g., `my-api.yaml` or `my-api.json`).                                                                                                                                                                                                                                                                                                                                         | Mandatory          |

> | `id`          | Id for the generation module                                                                                                                                                                                                                                                                                      | Mandatory        |

Reviewer 3: generation module or generated module?

Author: My mistake, it is a generated module.

| `targetModule`  | The Ballerina files are generated at the given directory name in the Ballerina package.                                                                                                                                                                                                                                                                                     | Optional          |

> | `options.mode`          | Mode type can be client. It generates client stubs for the given OpenAPI contract.                                                                                                                                                                                                                                                                                                 | Optional          |

Reviewer 3: ```suggestion
| `options.mode`          | The mode type can be `client`. It generates client stubs for the given OpenAPI contract.                                                                                                                                                                                                                                                                                                 | Optional          |
```

Reviewer 3: Shouldn't we mention the all possible mode types? Are they `client` and `service`?

Author: Yeah, we can add `service` mode also here. But currently, we only support client generation. Service mode will be supported in future releases. Therefore, I avoid including service mode at this time due to potential confusion for users with documentation. WDYT?

Reviewer 3: Then we better mention it or reword as `can be` is implying that there can be more other types. How about something like below?

The supported mode type is `client`. It generates client stubs for the given OpenAPI contract.

Author: This makes clear, that I will use this `The supported mode type is client. It generates client stubs for the given OpenAPI contract.` :)

| `options.tags`          | To generate the Ballerina client stub with a subset of tags defined in the OpenAPI contract                                                                                                                                                                                                             |  Optional        |
| `options.operations`    | To generate the Ballerina client stub with a subset of operations defined in the OpenAPI contract, use the `options.operations` option and specify the operations you need as specified in the OpenAPI definition.                                                                                                                                                                                                            |  Optional        |

> | `options.license`       | To generate the Ballerina files with the given copyright or license header, you can use this `options.license` option with path to your copyright text.                                                                                                                                                                                                                                                                                                         |  Optional         |

Reviewer 3: ```suggestion
| `options.license`       | To generate the Ballerina files with the given copyright or license header, you can use this `options.license` option with the path to your copyright text.                                                                                                                                                                                                                                                                                                         |  Optional         |
```

```
$ bal openapi add [-i | --input] <openapi-contract-file-path>
                  [--id] <client-id>
                  [--module] <target-module>
                  [--package] <ballerina-package-path>
```

For example,

```
$ bal openapi add -i <openapi-contract> --id <client_id> --module <target-module> --package <ballerina package path> --tags <"tag1","tag2">
```

| Command option      | Description                                                                                                                                                                                                                                                                                                                                                                     | Mandatory/Optional |
|----------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------|

> | `-i \| --input`   | The `-i|--input` command option specifies the path of the OpenAPI contract file (e.g., `my-api.yaml` or `my-api.json`).                                                                                                                                                                                                                                                                                                                                         | Mandatory          |

Reviewer 3: Incorrect table line content it seems. 

<img width="1091" alt="image" src="https://github.com/ballerina-platform/ballerina-dev-website/assets/16300038/41b41761-cc61-4c05-b59d-d7fe1034f2fd">

> | `-i \| --input`   | The `-i|--input` command option specifies the path of the OpenAPI contract file (e.g., `my-api.yaml` or `my-api.json`).                                                                                                                                                                                                                                                                                                                                         | Mandatory          |

Reviewer 4: The table seems to be broken due to this unscaped char. Please check.
```suggestion
| `-i \| --input`   | The `-i\|--input` command option specifies the path of the OpenAPI contract file (e.g., `my-api.yaml` or `my-api.json`).                                                                                                                                                                                                                                                                                                                                         | Mandatory          |
```

| `--id`          | Id for the generation module | Mandatory          |

> | `--module`          | Module name for the generated client stub module (defualt will be generated diractory)|  Optional        |

Reviewer 3: ```suggestion
| `--module`          | Module name for the generated client stub module (default will be generated directory)|  Optional        |
```
